packages/loopring_v3/contracts/amm/libamm/AmmExitRequest.sol

```diff
@@ -0,0 +1,174 @@
+            S, address(this), amounts[0], approvedByOperator
+        );
+
+        for (uint i = 0; i < size; i++) {
+            amountOuts[i + 1] = _withdrawToken(
+                S, S.tokens[i].addr, amounts[i + 1], approvedByOperator
+            );
+        }
+    }
+
+    function exitPool(
+        AmmData.State storage S,
+        uint                  poolAmountIn,
+        uint96[]     calldata minAmountsOut,
+        bool                  toLayer2
+        )
+        public
+        returns(AmmData.PoolExit memory exit)
+    {
+        require(minAmountsOut.length == S.tokens.length, "INVALID_DATA");
+
+        // To make the the available liqudity tokens cannot suddenly change
+        // we keep track of when onchain exits (which need to be processed) are pending.
+        require(S.isExiting[msg.sender] == false, "ALREADY_EXITING");
+        S.isExiting[msg.sender] = true;
+
+        exit = AmmData.PoolExit({
+            owner: msg.sender,
+            toLayer2: toLayer2,
+            poolAmountIn: poolAmountIn,
+            minAmountsOut: minAmountsOut,
+            storageIDs: new uint32[](0),
+            validUntil: 0xffffffff
+        });
+
+        // Approve the exit
+        bytes32 txHash = AmmUtil.hashPoolExit(S.domainSeperator, exit);
+        S.approvedTx[txHash] = block.timestamp;
+    }
+
+    function _checkOperatorApproval(
+        AmmData.State storage S,
+        uint[]       calldata amounts,
+        bytes        calldata signature, // signature from Exchange operator
+        uint                  validUntil
+        )
+        private
+        returns (bool)
+    {
+
+        // Check if we can withdraw without unlocking with an approval
+        // from the operator.
+        if (signature.length == 0) {
+            require(validUntil == 0, "INVALID_VALUE");
+            return false;
+        }
+
+        require(validUntil >= block.timestamp, 'SIGNATURE_EXPIRED');
+
+        bytes32 withdrawHash = EIP712.hashPacked(
+            S.domainSeperator,
+            keccak256(
+                abi.encode(
+                    WITHDRAW_TYPEHASH,
+                    msg.sender,
+                    keccak256(abi.encodePacked(amounts)),
+                    validUntil,
+                    S.nonces[msg.sender]++
+                )
+            )
+        );
+        require(
+            withdrawHash.verifySignature(S.exchange.owner(), signature),
+            "INVALID_SIGNATURE"
+        );
+        return true;
+    }
+
+    function _withdrawToken(
+        AmmData.State storage S,
+        address               token,
+        uint                  amount,
+        bool                  approvedByOperator
+        )
+        private
+        returns (uint withdrawn)
+    {
+        uint available = approvedByOperator ?
+            S.lockedBalance[token][msg.sender] :
+            S.availableBalance(token, msg.sender);
+
+        withdrawn = (amount > available || amount == 0) ? available : amount;
```

Comment:
I believe `amount == 0` should be valid and not withdraw anything, otherwise we force people to always withdraw from all tokens which is strange? (and not something we actually want in some cases)

Reply:
OK